crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3083,12 +3117,100 @@ impl<'db> IterationOutcome<'db> {
         }
     }
 
-    fn unwrap_without_diagnostic(self) -> Type<'db> {
+    /// Returns the element type if it is known, or `None` if the type is never iterable.
+    fn element_type(&self) -> Option<Type<'db>> {
+        match self {
+            IterateError::NotIterable { .. } => None,
+            IterateError::PossiblyUnbound { element_ty, .. } => Some(*element_ty),
+        }
+    }
+
+    /// Returns the element type if it is known, or `Type::unknown()` if it is not.
+    fn unwrap_element_type(&self) -> Type<'db> {
+        self.element_type().unwrap_or(Type::unknown())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) enum BoolError<'db> {
+    /// The type has a `__bool__` attribute but it can't be called.
+    NotCallable { not_boolable_type: Type<'db> },
+
+    /// The type has a `__bool__` attribute, but it doesn't have the right signature.
+    IncorrectSignature {
+        not_boolable_type: Type<'db>,
+        return_type: Type<'db>,
```

Comment:
Creating a multi-span diagnostic that shows the return type probably requires us to return an entire `binding` (or `CallOutcome`) here.